aspnetcore/blazor/Security-guidance.md

```diff
@@ -0,0 +1,254 @@
+  * Maximum amount of unacknowledged render batches the server keeps in memory per circuit at a given time to support robust reconnection. It can be configured in `CircuitOptions.MaxBufferedUnacknowledgedRenderBatches` and after reaching this limit the server stops producing new render batches until one or more batches have been ack by the client.
+* Blazor server-side applications rely on the following SignalR and ASP.NET Core limits:
+  * Message size for an individual message. This is covered by the max message size in SignalR which defaults to 32Kb.
+
+### Additional protection
+By default there's no limit on the number of connections per client/user to a blazor server-side application. If your application requires such limit there are several approaches that can be taken.
+* Require authentication to connect to the application. This will naturally limit the ability of unauthorized users to connect to the application provided they can't provision a user for the system.
+* Limit the number of connections per user. This could be done at the application level through endpoint routing extensibility, by requiring authentication to connect to the app and keeping track of the active sessions per client/user and rejecting new sessions upon hitting a limit.
+* Limiting the number of connections per client/user can also be achieved by using a proxy/gateway in front of your application like Azure CloudFront.
+* In the last two cases, care needs to be taken to prevent denying access to legitimate users, if for example the limit is established based on the client ip.
+
+## Authentication and authorization
+* For specific guidance on authentication and authorization check the docs at https://docs.microsoft.com/en-us/aspnet/core/security/blazor/?view=aspnetcore-3.0
+
+## Security check-list for blazor server-side applications
+What follows is a **basic and non exhaustive** list of checks and recommendations to follow as part of considering the security aspects of your blazor server-side application.
+* Validate arguments from events.
+* Validate results from JS interop calls.
+* Validate inputs from JS interop calls.
+* Avoid using or validate beforehand user input for .NET to JS interop calls.
+* Prevent the client from allocating an unbound amount of memory.
+  * Data within the component.
+  * DotNetObject references returned to the client.
+* Avoid using user input as part of calls to NavigationManager.Navigate and validate it against a set of allowed origins first if unavoidable.
+* Do not make authorization decissions based on the state of the UI but only from state from the component.
+* Consider using CSP to protect agaisnt XSS attacks.
+* Consider using CSP and X-Frame-Options to protect against click-jacking.
+* Ensure CORS settings are acceptable when enabling CORS or explicitly disable it for Blazor.
+* Avoid using script tags in components.
```

Comment:
Probably don't need this one since it's a built-in rule in the compiler (and for anyone writing RenderTreeBuilder logic by hand, they already know they are doing dangerous stuff with bigger hazards than just producing script tags).